helix-vcs/src/differ.rs

```diff
@@ -0,0 +1,164 @@
+use std::mem::take;
+use std::ops::{Deref, Range};
+use std::sync::Arc;
+
+use arc_swap::ArcSwap;
+use helix_core::{Rope, RopeSlice};
+use imara_diff::intern::InternedInput;
+use imara_diff::Algorithm;
+use tokio::sync::mpsc::{unbounded_channel, UnboundedReceiver, UnboundedSender};
+use tokio::task::JoinHandle;
+use tokio::time::{timeout, Duration};
+
+use crate::rope_line_cache::InternedRopeLines;
+use crate::{LineDiff, LineDiffs};
+
+#[cfg(test)]
+mod test;
+
+#[derive(Clone, Debug)]
+pub struct Differ {
+    channel: UnboundedSender<Event>,
+    line_diffs: Arc<ArcSwap<LineDiffs>>,
+}
+
+impl Differ {
+    pub fn new(diff_base: Rope, doc: Rope) -> Differ {
+        Differ::new_with_handle(diff_base, doc).0
+    }
+
+    fn new_with_handle(diff_base: Rope, doc: Rope) -> (Differ, JoinHandle<()>) {
+        let (sender, receiver) = unbounded_channel();
```

Comment:
Is it a good idea to use unbounded channel instead of bounded channel?

Reply:
This is fine because we have a de-bouncer. If the differ falls behind we simply merge all events from the channel before we perform the diff. That means the only way an unbounded channel could overflow here is if there were enough events to exhaust system memory while the diff is performed. There is a fairly conservative bound (~65k lines or 65k*128) on files size after which the diff simply does nothing. At these sizes `imara-diff` is pretty fast (at most 200ms, likely much much faster). That means helix would need to generate enough events to overflow a channel in 200ms or less. 

On the other hand a bounded channel would cause the main thread to wait on the diff in case the buffer is filled up which I think is strictly a worse.